### PR TITLE
Rename ResilienceStrategy to ResiliencePipeline

### PR DESCRIPTION
**Pull request by martintmk**
### Details on the issue fix or feature implementation

Based on the API review:
https://github.com/App-vNext/Polly/pull/1233#issuecomment-1673200949

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

**Comment by peter-csala** [comment]
@martintmk Did I miss something? I though you are against the usage of the term *pipeline*?

**Comment by martintmk** [comment]
> @martintmk Did I miss something? I though you are against the usage of the term _pipeline_?

We did some brainstorming and realized we need to differentiate between "pipeline" (was strategy before) and a component of a pipeline (individual resilience strategy).

The end consumer (majority) of an API only cares about `ResiliencePipeline`. The `ResiliencePipeline` can only ever be created by using `ResiliencePipelineBuilder`. Its creation and state is fully under our control, so both `ResiliencePipeline` and `ResiliencePipeline<T>` is sealed and with internal constructor. This allows us to evolve and optimize the internal implementations without breaking changes.

The resilience author (minority) cares about `ResilienceStrategy` and `ResilienceStrategy<T>`, they have simple and straightforward base class to implement followed by exposing extensions for `ResiliencePipelineBuilder`.

File: bench/Polly.Core.Benchmarks/CircuitBreakerOpenedBenchmark.cs [hunk 1]
```diff
@@ -2,15 +2,15 @@ namespace Polly.Core.Benchmarks;
 
 public class CircuitBreakerOpenedBenchmark
 {
-    private ResilienceStrategy? _strategy;
-    private ResilienceStrategy? _strategyHandlesOutcome;
+    private ResiliencePipeline? _strategy;
```
Comment (martincostello) [inline bench/Polly.Core.Benchmarks/CircuitBreakerOpenedBenchmark.cs:5]: Should we also rename all these locals and fields?

File: src/Polly.Core/CircuitBreaker/CircuitBreakerResiliencePipelineBuilderExtensions.cs [hunk 2]
```diff
@@ -6,9 +6,9 @@
 namespace Polly;
 
 /// <summary>
-/// Circuit breaker strategy extensions for <see cref="CompositeStrategyBuilder"/>.
+/// Circuit breaker strategy extensions for <see cref="ResiliencePipelineBuilder"/>.
```
Comment (martincostello) [inline src/Polly.Core/CircuitBreaker/CircuitBreakerResiliencePipelineBuilderExtensions.cs:9]: Similarly uses of strategy in the comments.

File: src/Polly.Core/Hedging/HedgingResiliencePipelineBuilderExtensions.cs [hunk 3]
```diff
@@ -6,15 +6,15 @@
 namespace Polly;
 
 /// <summary>
-/// Provides extension methods for configuring hedging resilience strategies for <see cref="CompositeStrategyBuilder"/>.
+/// Provides extension methods for configuring hedging resilience strategies for <see cref="ResiliencePipelineBuilder"/>.
```
Comment (martincostello) [inline src/Polly.Core/Hedging/HedgingResiliencePipelineBuilderExtensions.cs:9]: Same for "strategies" too.

File: src/Polly.Core/NullResiliencePipeline.TResult.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+namespace Polly;
+
+/// <summary>
+/// A resilience pipeline that executes an user-provided callback without any additional logic.
+/// </summary>
+/// <typeparam name="TResult">The type of result this strategy handles.</typeparam>
```
Comment (martincostello) [inline src/Polly.Core/NullResiliencePipeline.TResult.cs:6]:
```suggestion
/// <typeparam name="TResult">The type of result this pipeline handles.</typeparam>
```

File: src/Polly.Core/Registry/ConfigureBuilderContext.cs [hunk 5]
```diff
@@ -3,23 +3,23 @@
...
     internal ConfigureBuilderContext(TKey strategyKey, string builderName, string? builderInstanceName)
     {
-        StrategyKey = strategyKey;
+        PipelineKey = strategyKey;
         BuilderName = builderName;
         BuilderInstanceName = builderInstanceName;
     }
 
     /// <summary>
     /// Gets the strategy key for the strategy being created.
```
Comment (martincostello) [inline src/Polly.Core/Registry/ConfigureBuilderContext.cs:20]: Pipeline.

File: src/Polly.Core/Telemetry/ResilienceTelemetrySource.cs [hunk 6]
```diff
@@ -11,31 +11,31 @@ public sealed class ResilienceTelemetrySource
     /// <summary>
     /// Initializes a new instance of the <see cref="ResilienceTelemetrySource"/> class.
     /// </summary>
-    /// <param name="builderName">The builder name.</param>
-    /// <param name="builderInstanceName">The builder instance name.</param>
+    /// <param name="pipelineName">The builder name.</param>
+    /// <param name="pipelineInstanceName">The builder instance name.</param>
```
Comment (martincostello) [inline src/Polly.Core/Telemetry/ResilienceTelemetrySource.cs:15]:
```suggestion
    /// <param name="pipelineName">The pipeline name.</param>
    /// <param name="pipelineInstanceName">The pipeline instance name.</param>
```

File: src/Polly.Core/Telemetry/ResilienceTelemetrySource.cs [hunk 7]
```diff
@@ -11,31 +11,31 @@ public sealed class ResilienceTelemetrySource
...
-        BuilderName = builderName;
-        BuilderInstanceName = builderInstanceName;
+        PipelineName = pipelineName;
+        PipelineInstanceName = pipelineInstanceName;
         BuilderProperties = builderProperties;
         StrategyName = strategyName;
     }
 
     /// <summary>
     /// Gets the builder name.
```
Comment (martincostello) [inline src/Polly.Core/Telemetry/ResilienceTelemetrySource.cs:31]: Pipeline

File: src/Polly.Core/Telemetry/ResilienceTelemetrySource.cs [hunk 8]
```diff
@@ -11,31 +11,31 @@ public sealed class ResilienceTelemetrySource
...
     /// </summary>
-    public string? BuilderName { get; }
+    public string? PipelineName { get; }
 
     /// <summary>
     /// Gets the builder instance name.
```
Comment (martincostello) [inline src/Polly.Core/Telemetry/ResilienceTelemetrySource.cs:36]: Pipeline

**Review by martincostello — APPROVED** [review]

**Pull request merged**